strawberryfields/tdm/tdmprogram.py

```diff
@@ -303,6 +304,98 @@ def context(self, *args, copies=1, shift="default"):
+        At this stage the compilation is simply a check that the program matches the device.
+
+        Args:
+            device (~strawberryfields.api.DeviceSpec): device specification object to use for
+                program compilation
+            compiler (str, ~strawberryfields.compilers.Compiler): Compiler name or compile strategy
+                to use. If a device is specified, this overrides the compile strategy specified by
+                the hardware :class:`~.DeviceSpec`. For :class:`~.TDMProgram` this should be set to
+                None.
+
+        Returns:
+            Program: compiled program
+        """
+        if compiler is not None:
+            raise CircuitError("Program can only be compiled with default compiler")
+
+        device_layout = bb.loads(device.layout)
+
+        # First check: the gates are in the correct order
+        program_gates = [cmd.op.__class__.__name__ for cmd in self.rolled_circuit]
+        device_gates = [op["op"] for op in device_layout.operations]
+        if device_gates != program_gates:
+            raise CircuitError(
+                "Program cannot be used with the device '{}' "
+                "due to incompatible topology.".format(device.target)
+            )
+
+        # Second check: the gates act in the correct modes
+        program_modes = [[r.ind for r in cmd.reg] for cmd in self.rolled_circuit]
+        device_modes = [op["modes"] for op in device_layout.operations]
+        if program_modes != device_modes:
+            raise CircuitError(
+                "Program cannot be used with the device '{}' "
+                "due to incompatible mode ordering.".format(device.target)
+            )
+
+        # Third check: the parameters of the gates are valid
+        gate_params_ranges = device.gate_parameters
+        # We will loop over the different operations in the device specification
+
+        for i, operation in enumerate(device_layout.operations):
+            # We obtain the name of the parameter(s)
+            param_names = operation["args"]
+            len_params_program = len(self.rolled_circuit[i].op.p)
+            len_params_device = len(param_names)
+            # The next if is to make sure we do not flag incorrectly things like Sgate(r,0) beign different Sgate(r)
+
+            if len_params_device < len_params_program:
+                for j in range(1, len_params_program):
+                    if self.rolled_circuit[i].op.p[j] != 0:
+                        raise CircuitError(
+                            "Program cannot be used with the device '{}' "
+                            "due to incompatible parameter.".format(device.target)
+                        )
+            # Now we will check explicitly if the parameters in the program match
+            counter = 0  # counts the number of symbolic variables, which are labeled consecutively by the context method
```

Comment:
(Minor suggestion): how about `num_symbolic_param`/`num_symbolic` as a name?

Reply:
Thanks, yes, it is much more descriptive.